README.md
It's a simple countdown class in JavaScript. What you can do with it? Basically there are two modes 'date' and 'timer'. First one will measure time between dates, second one, will measure time from some value in miliseconds to zero.

### Creating object:
- 'timer' mode:
```javascript       
        var settings = {
            mode : 'timer', //required!!
            limit : number, //time limit in ms, required!!
            pattern : string, //how should presentation of countdown look like, default 
            //'%D days %HH hours %MM minutes %SS seconds %UU miliseconds'
    
            //%D - days from 0 to n
            //%HH, %H - hours
            //%MM, %M - minutes
            //%SS, %S - seconds
            //%UU, %U - miliseconds
    
            //if something has double letter, it means that 0's will be put before, e.g. 12:04:02 instead of 12:4:2
    
            nodeId : , //string, required, pattern will be placed in this node
            speed : 1, //default 1,
            refreshRate, //number, default 100 ms, how often counter should be refreshed
            inactiveTab, //boolean, default true, should countdown work when tab is inactive
    
            //user events, and the first argument is current object
            onStop : function(countdownObject){}, 
            onReset : function(countdownObject){},
            onStart : function(countdownObject){},
            onTick : function(countdownObject){},
            onEnd : function(countdownObject){}
        };

        var ob = new Countdown(settings);
```
   

- 'date' mode:
```javascript
    var settings = {
        mode : 'date',
        startDate : , //Date object, default new Date(),
        endDate : , //Date object required!!, 
        ...
    };
```

There is no type checking so be careful with types.

-----------------------------------------------------------------------------

### On objects you can call:
- getTimerId() //returns interval id
- getState() //returns current state init, started, stopped, ended
- getTime()
  
  returns object with current time 

        {
            days : ,
            hours : ,
            minutes : ,
            seconds : ,
            miliseconds : ,
        } 

- start()
- stop()
- reset() //avaiable only for 'timer' mode
- setSpeed(number) //you can adjust speed, make it faster or slower
 
### Example:
http://jsfiddle.net/Rymar/H8fxU/

CHANGES.md
v0.1 -> 2012.08.25
----------------------
Initial version

countdown-min.js
var Countdown=function(){var b=function(a,c){for(var e=c-(""+a).length,b="",d=0;d<e;d++)b+="0";return b+a},d=function(a){this._miliseconds=this._seconds=this._minutes=this._hours=this._days=this._timerId=0;this._endDate=this._startDate=null;this._pattern="%D days %HH hours %MM minutes %SS seconds %UU miliseconds";this._mode="";this._timeLeft=this._node=null;this._limit=0;this._refreshRate=10;this._speed=1;this._state="init";this._inactiveTab=!0;this._before=null;this._settings=a;var c=0,b;for(b="date"===
a.mode?["mode","endDate","nodeId"]:["mode","limit","nodeId"];c<b.length;c++)if(!(b[c]in a))return console.error(b[c]+" is required"),!1;this._mode=a.mode;this._inactiveTab=a.inactiveTab||!0;this._speed=a.speed||1;this._refreshRate=(a.refreshRate&&10>a.refreshRate?10:a.refreshRate)||100;this._pattern=a.pattern||this._pattern;this._limit=a.limit;this._endDate=a.endDate;this._node=document.getElementById(a.nodeId)};d.prototype={getTimerId:function(){return this._timerId},getTime:function(){return{days:this._days,
hours:this._hours,minutes:this._minutes,seconds:this._seconds,miliseconds:this._miliseconds}},getState:function(){return this._state},setSpeed:function(a){this._speed=a},_tick:function(a){var b=+new Date-this._before,a="undefined"!==typeof a?a:this._refreshRate;this._timeLeft=this._inactiveTab&&b>a+10?this._timeLeft-b*this._speed:this._timeLeft-a*this._speed;0>=this._timeLeft?this._end():(this._count(),this._show(),this._settings.onTick&&this._settings.onTick(this),this._before=new Date)},start:function(){"date"===
this._mode?(this._startDate=this._settings.startDate||new Date,this._timeLeft=this._endDate-this._startDate):"timer"===this._mode&&null===this._timeLeft&&(this._timeLeft=this._limit);this._before=new Date;"init"===this._state&&this._tick(0);this._state="started";var a=this;this._timerId=setInterval(function(){a._tick(this._refreshRate)},this._refreshRate);this._settings.onStart&&this._settings.onStart(this);0>=this._timeLeft&&this._end()},_count:function(){this._miliseconds=this._timeLeft%1E3/10|
0;this._seconds=(this._timeLeft/1E3|0)%60;this._minutes=(this._timeLeft/6E4|0)%60;this._hours=(this._timeLeft/36E5|0)%24;this._days=this._timeLeft/864E5|0},_show:function(){var a=this._pattern,a=a.replace(/%HH/g,b(this._hours,2)),a=a.replace(/%MM/g,b(this._minutes,2)),a=a.replace(/%SS/g,b(this._seconds,2)),a=a.replace(/%UU/g,b(this._miliseconds,2)),a=a.replace(/%D/g,this._days),a=a.replace(/%H/g,this._hours),a=a.replace(/%M/g,this._minutes),a=a.replace(/%S/g,this._seconds),a=a.replace(/%U/g,this._miliseconds);
this._node.innerHTML=a},stop:function(){clearInterval(this._timerId);this._state="stopped";this._settings.onStop&&this._settings.onStop(this)},_end:function(){this._days=this._hours=this._minutes=this._seconds=this._miliseconds=0;this._show();clearInterval(this._timerId);this._state="ended";this._settings.onEnd&&this._settings.onEnd(this)},reset:function(){"date"!==this._mode&&(this._state=null,this._timeLeft=this._limit,this._count(),this._show(),this.stop(),this._settings.onReset&&this._settings.onReset(this))}};
return function(a){return new d(a)}}();

countdown.js

/**
 * @author Mariusz Rymarczyk
 * @license MIT
 * @version 0.1
 */

'use strict';

var Countdown = (function () {

    /**
     * Adding preceding zeros
     * @param val int 
     * @param n int ciphers in number
     */
    var fullFormat = function (val, n) {

        var len = ('' + val).length,
            zeros = n - len,
            tmp = '',
            i = 0;

        for ( ; i < zeros ; i++) {
            tmp += '0';
        }

        return tmp + val;
    };

    var Countdown = function (settings) {

        /* pseudo private */

        /* setInterval id */
        this._timerId = 0;
        /* days left */
        this._days = 0;
        /* hours left */
        this._hours = 0;
        /* minutes left */
        this._minutes = 0;
        /* seconds left */
        this._seconds = 0;
        /* miliseconds left */
        this._miliseconds = 0;
        /* start date Object */
        this._startDate = null;
        /* end date Object */
        this._endDate = null;
        /* how the date will be showed on page */
        this._pattern = '%D days %HH hours %MM minutes %SS seconds %UU miliseconds';
        /* mode of countdown: 'date' or 'timer' */ /*REQUIRED!!*/
        this._mode = '';
        /* element where remaining time will be placed */ /*REQUIRED!!*/
        this._node = null;
        /* in seconds or miliseconds, depends on settings */
        this._timeLeft = null;
        /* time limit for timer mode */
        this._limit = 0;
        /* step of interval */
        this._refreshRate = 10;
        /* speed ratio*/
        this._speed = 1;
        /* state */
        this._state = 'init';
        /* inactiveTab */
        this._inactiveTab = true;
        /* date before changing tab */
        this._before = null;
        /* initial settings */
        this._settings = settings;

        /* private */
        var /* iteration var */
            i = 0,
            /* required settings */
            required;

        required = settings.mode === 'date' ? ['mode', 'endDate', 'nodeId'] : ['mode', 'limit', 'nodeId'];

        /* check required fields */
        for ( ; i < required.length ; i++) {

            if (!(required[i] in settings)) {

                console.error(required[i] + ' is required');
                return false;
            }
        }

        /* initialization */

        this._mode = settings.mode;

        this._inactiveTab = settings.inactiveTab || true;

        this._speed = settings.speed || 1;

        this._refreshRate = ((settings.refreshRate && settings.refreshRate < 10) ? 10 : settings.refreshRate) || 100;

        this._pattern = settings.pattern || this._pattern;
        
        this._limit = settings.limit;

        this._endDate = settings.endDate;
        
        this._node = document.getElementById(settings.nodeId);

        
        /*var ret = {
            getTimerId : getTimerId,
            getTime : getTime,
            getState : getState,
            start : start,
            stop : stop,
            setSpeed : setSpeed
        };*/

        /*if(mode === 'timer') {
            ret.reset = reset;
        }*/
    };

    Countdown.prototype = {

        getTimerId : function () { 
            return this._timerId; 
        },

        getTime : function () {

            return {
                days : this._days,
                hours : this._hours,
                minutes : this._minutes,
                seconds : this._seconds,
                miliseconds : this._miliseconds
            };

        },

        getState : function () {
            return this._state;
        },

        setSpeed : function (val) {
            this._speed = val;
        },

        /**
         * Executed by setInterval
         * @param timeStep number not required
         */
        _tick : function (timeStep) {

            var elapsedTime = +new Date() - this._before,
                tStep = typeof timeStep !== 'undefined' ? timeStep : this._refreshRate;

            //fixing inactive tab time
            if (this._inactiveTab && (elapsedTime > tStep + 10) ) {
                this._timeLeft -= elapsedTime * this._speed;
            } else {
                this._timeLeft -= tStep * this._speed; 
            }

            if(this._timeLeft <= 0) {
                this._end();
                return;
            }

            this._count();
            this._show();
            
            this._settings.onTick && this._settings.onTick(this);
           
            this._before = new Date();
        },

        /* start counting */
        start : function () {

            if (this._mode === 'date') {

                this._startDate = this._settings.startDate || new Date();

                this._timeLeft = this._endDate - this._startDate;

            } else if (this._mode === 'timer') {

                if (this._timeLeft === null) {
                    this._timeLeft = this._limit;
                }
            }

            this._before = new Date();

            if(this._state === 'init'){
                this._tick(0);
            }

            this._state = 'started';

            var self = this;

            this._timerId = setInterval(function(){
                self._tick(this._refreshRate);
            }, this._refreshRate);

            this._settings.onStart && this._settings.onStart(this);

            if(this._timeLeft <= 0){
                this._end();
            }
        },

        _count : function () {

            this._miliseconds = (this._timeLeft%1000)/10 | 0;
            this._seconds = ((this._timeLeft/1000) | 0) % 60;
            this._minutes = ((this._timeLeft/60000) | 0) % 60;
            this._hours = ((this._timeLeft/3600000) | 0) % 24;
            this._days = (this._timeLeft/86400000) | 0;

        },

        _show : function () {
            
            var content = this._pattern;

            content = content.replace(/%HH/g, fullFormat(this._hours, 2));
            content = content.replace(/%MM/g, fullFormat(this._minutes, 2));
            content = content.replace(/%SS/g, fullFormat(this._seconds, 2));
            content = content.replace(/%UU/g, fullFormat(this._miliseconds, 2));

            content = content.replace(/%D/g, this._days);
            content = content.replace(/%H/g, this._hours);
            content = content.replace(/%M/g, this._minutes);
            content = content.replace(/%S/g, this._seconds);
            content = content.replace(/%U/g, this._miliseconds);
          
            this._node.innerHTML = content;
        },

        /* stop counting */
        stop : function () {
            clearInterval(this._timerId);
            this._state = 'stopped';
            this._settings.onStop && this._settings.onStop(this);
        },

        /* end */
        _end : function () {
            this._days = this._hours = this._minutes = this._seconds = this._miliseconds = 0;
            this._show();
            clearInterval(this._timerId);
            this._state = 'ended';
            this._settings.onEnd && this._settings.onEnd(this);
        },

        /* reset exists only for timer mode, reset to original time */
        reset : function () {

            if(this._mode === 'date'){ return; }
            
            this._state = null;
            this._timeLeft = this._limit;
            this._count();
            this._show();
            this.stop();
            this._settings.onReset && this._settings.onReset(this);
        }
    }

    return function (settings) {
        return new Countdown(settings);
    };

})();